subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/devices/provider/KubernetesDevicesProvider.kt

```diff
@@ -0,0 +1,29 @@
+package com.avito.instrumentation.reservation.devices.provider
+
+import com.avito.instrumentation.reservation.client.kubernetes.KubernetesReservationClient
+import com.avito.instrumentation.reservation.request.Reservation
+import com.avito.runner.service.worker.device.Device
+import com.avito.runner.service.worker.device.adb.AdbDevicesManager
+import com.avito.utils.logging.CILogger
+import kotlinx.coroutines.channels.ReceiveChannel
+import kotlinx.coroutines.channels.map
+
+class KubernetesDevicesProvider(
+  private val client: KubernetesReservationClient,
+  private val adbDevicesManager: AdbDevicesManager,
+  private val logger: CILogger
+) : DevicesProvider {
+
+    override suspend fun provideFor(reservations: Collection<Reservation.Data>): ReceiveChannel<Device> {
+        val claim = client.claim(reservations)
+        // TODO parallel device getting
+        return claim.serials.map { serial ->
+            adbDevicesManager.findDevice(serial)
+                .orElseGet { throw IllegalStateException("Can't find device connected adb device $serial") }
```

Comment:
What will it fail? Only the one reading from the channel?

Reply:
This code is moved from deleted class `ExternalIntentionBasedDevicesObserver`

Reply:
It will break the whole task

Reply:
> It will break the whole task

This is strange. If one emulator hangs why the whole task should fail? 
It looks like we should restart this one like from DeviceWorker